sklearn/preprocessing/_data.py

```diff
@@ -1802,6 +1806,342 @@ def transform(self, X):
+        self.bsplines_ = np.array(bsplines)
+
+        self.n_input_features_ = n_features
+        self.n_output_features_ = n_out - n_features * self.include_bias
+        return self
+
+    def transform(self, X):
+        """Transform each feature data to B-splines.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data to transform.
+
+        Returns
+        -------
+        XBS : ndarray of shape (n_samples, n_features * n_splines)
+            The matrix of features, where n_splines is the number of bases
+            elements of the B-splines, n_knots + degree - 1.
+        """
+        check_is_fitted(self)
+
+        X = check_array(X)
+
+        n_samples, n_features = X.shape
+        n_splines = self.bsplines_[0].c.shape[0]
+        degree = self.degree
+
+        if n_features != self.n_input_features_:
+            raise ValueError("X shape does not match training shape")
+
+        # Note that scipy splines return float64.
+        n_out = self.n_output_features_ + n_features * self.include_bias
+        XBS = np.zeros((n_samples, n_out), dtype=np.float64,
+                       order=self.order)
+
+        for i in range(n_features):
+            spl = self.bsplines_[i]
+
+            if self.extrapolation in ('continue', 'error'):
+                XBS[:, (i * n_splines):((i + 1) * n_splines)] = spl(X[:, i])
+            else:
+                xmin = spl.t[degree]
+                xmax = spl.t[-degree - 1]
+                mask = (xmin <= X[:, i]) & (X[:, i] <= xmax)
+                XBS[mask, (i * n_splines):((i + 1) * n_splines)] = (
+                    spl(X[mask, i]))
+
+            # Note for extrapolation:
+            # 'continue' is already returned as is by scipy BSplines
+            if self.extrapolation == 'error':
+                # BSpline with extrapolate=False does not raise an error, but
+                # output np.nan.
+                assert_all_finite(
+                    XBS[:, (i * n_splines):((i + 1) * n_splines)])
+            elif self.extrapolation == 'constant':
+                # Set all values beyond xmin and xmax to the value of the
+                # spline basis functions at those two positions.
+                # Only the first degree and last degree number of splines
+                # have non-zero values at the boundaries.
+
+                # spline values at boundaries
+                f_min = spl(xmin)
+                f_max = spl(xmax)
+                mask = X[:, i] < xmin
+                if np.any(mask):
+                    XBS[mask, (i * n_splines):(i * n_splines + degree)] = (
+                        f_min[:degree])
+
+                mask = X[:, i] > xmax
+                if np.any(mask):
+                    XBS[mask,
+                        ((i+1) * n_splines - degree):((i+1) * n_splines)] = (
+                        f_max[-degree:])
+            elif self.extrapolation == 'linear':
```

Comment:
Is there a reference that suggest that `extrapolation='linear'` is would be correct?

Reply:
See docs of the R package [mgcv](https://cran.r-project.org/web/packages/mgcv/mgcv.pdf):
> Linear extrapolation is used for prediction that requires extrapolation (i.e.  prediction outside therange of the interiork-m[1]+1knots — the interval over which the penalty is evaluated).  Suchextrapolation is not allowed in basis construction, but is when predicting.